moveit_core/planning_scene/src/planning_scene.cpp

```diff
@@ -411,58 +417,146 @@ void PlanningScene::checkCollision(const collision_detection::CollisionRequest&
   checkCollision(req, res, robot_state, getAllowedCollisionMatrix());
 }
 
-void PlanningScene::checkSelfCollision(const collision_detection::CollisionRequest& req,
-                                       collision_detection::CollisionResult& res)
+void PlanningScene::checkCollision(const collision_detection::CollisionRequest& req,
+                                   collision_detection::CollisionResult& res, moveit::core::RobotState& robot_state,
+                                   const collision_detection::AllowedCollisionMatrix& acm) const
 {
-  if (getCurrentState().dirtyCollisionBodyTransforms())
+  if (robot_state.dirtyCollisionBodyTransforms())
   {
-    checkSelfCollision(req, res, getCurrentStateNonConst());
-  }
-  else
-  {
-    checkSelfCollision(req, res, getCurrentState());
+    robot_state.updateCollisionBodyTransforms();
   }
+  checkCollision(req, res, static_cast<const moveit::core::RobotState&>(robot_state), acm);
 }
 
 void PlanningScene::checkCollision(const collision_detection::CollisionRequest& req,
                                    collision_detection::CollisionResult& res,
                                    const moveit::core::RobotState& robot_state,
-                                   const collision_detection::AllowedCollisionMatrix& acm) const
+                                   const collision_detection::AllowedCollisionMatrix& acm,
+                                   bool validate_transforms) const
 {
+  if (validate_transforms && robot_state.dirtyCollisionBodyTransforms())
```

Comment:
Do we really need `validate_transforms` if it's just to show a warning? i.e. it's not making this function fail in any way.
And `dirtyCollisionBodyTransforms` should be a cheap check anyway since it's just checking an internal member variable of the RobotState.

Reply:
I don't like about this function that you could silently do the wrong thing if you're not update the collision transforms :thinking: However, I wanted to give the option to disable the warning with this flag (e.g. if you run this in a fast loop). On second thought, it might be a better idea to warn on default and the user needs to actively disable the warning. Alternatively,  we could do a bigger change and return an error message with the collision result. I am hesitant to do this in this PR though to get this change in first